File: source/328.奇偶链表.cpp
/*
 * @lc app=leetcode.cn id=328 lang=cpp
 *
 * [328] 奇偶链表
 */
#include "cpp_includes.h"
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (!head) return head;

        auto odd = head;
        auto eve = head->next;
        auto eveHead = head->next;

        while(eve && eve->next) {
            odd->next = eve->next;
            odd = odd->next;
            eve->next = odd->next;
            eve = eve->next;
        }

        odd->next = eveHead;
        return head;
    }
};
// @lc code=end

File: source/328.奇偶链表.cs
/*
 * @lc app=leetcode.cn id=328 lang=csharp
 *
 * [328] 奇偶链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
  public ListNode OddEvenList(ListNode head)
  {
    if (head == null || head.next == null) return head;

    var odd = head;
    var even = head.next;
    var evenHead = head.next;

    while (even != null && even.next != null)
    {
      odd.next = even.next;
      odd = odd.next;
      even.next = odd.next;
      even = even.next;
    }

    odd.next = evenHead;
    return head;
  }
}
// @lc code=end

